samza-core/src/main/java/org/apache/samza/container/grouper/stream/AllSspToSingleTaskGrouperFactory.java

```diff
@@ -19,51 +19,116 @@
+    Map<TaskName, Set<SystemStreamPartition>> groupedMap = new HashMap<>();
+
     if (ssps == null) {
       throw new SamzaException("ssp set cannot be null!");
     }
     if (ssps.size() == 0) {
       throw new SamzaException("Cannot process stream task with no input system stream partitions");
     }
 
-    final TaskName taskName = new TaskName(String.format("Task-%s", String.valueOf(containerId)));
+    if (isPassthroughJobCoordinator) {
+      final TaskName taskName = new TaskName(String.format("Task-%d", containerId));
+      groupedMap.put(taskName, ssps);
+    } else {
+      // Assign all partitions to each task
+      IntStream.range(0, taskCount).forEach(i -> {
+          final TaskName taskName = new TaskName(String.format("Task-%s", String.valueOf(i)));
+          groupedMap.put(taskName, ssps);
+        });
+    }
 
-    return Collections.singletonMap(taskName, ssps);
+    return groupedMap;
   }
 }
 
 public class AllSspToSingleTaskGrouperFactory implements SystemStreamPartitionGrouperFactory {
+  private static final String YARN_CONTAINER_COUNT = "yarn.container.count";
+  private static final String PASSTHROUGH_JOB_COORDINATOR_FACTORY = PassthroughJobCoordinatorFactory.class.getName();
+  private static final String ZK_JOB_COORDINATOR_FACTORY = ZkJobCoordinatorFactory.class.getName();
+
   @Override
   public SystemStreamPartitionGrouper getSystemStreamPartitionGrouper(Config config) {
-    return new AllSspToSingleTaskGrouper(config.getInt(JobConfig.PROCESSOR_ID()));
+    int taskCount;
+    int containerId = 0;
+    boolean isPassthroughJobCoordinator = false;
+    String jobCoordinatorFactoryClassName = config.get(JobCoordinatorConfig.JOB_COORDINATOR_FACTORY);
+
+    if (jobCoordinatorFactoryClassName != null && !jobCoordinatorFactoryClassName.isEmpty()) {
+      if (jobCoordinatorFactoryClassName.equals(ZK_JOB_COORDINATOR_FACTORY)) {
+        taskCount = getNumProcessors(config);
+      } else if (jobCoordinatorFactoryClassName.equals(PASSTHROUGH_JOB_COORDINATOR_FACTORY)) {
+        if (config.get(JobConfig.PROCESSOR_ID()) == null) {
+          throw new ConfigException("Could not find " + JobConfig.PROCESSOR_ID() + " in Config!");
+        }
+        isPassthroughJobCoordinator = true;
+        containerId = config.getInt(JobConfig.PROCESSOR_ID());
+        taskCount = 1;
+      } else {
+        throw new ConfigException("AllSspToSingleTaskGrouperFactory is not yet supported with "
+            + jobCoordinatorFactoryClassName);
+      }
+    } else {
+      // must be cluster-based jobCoordinator (Yarn)
+      taskCount = config.getInt(YARN_CONTAINER_COUNT);
+    }
+
+    Validate.isTrue(taskCount > 0, "Task count cannot be <= 0");
+
+    return new AllSspToSingleTaskGrouper(isPassthroughJobCoordinator, containerId, taskCount);
+  }
+
+  private int getNumProcessors(Config config) {
+    ZkConfig zkConfig = new ZkConfig(config);
```

Comment:
This puts a requirement that when getSystemStreamPartitionGrouper() is called, the ZkJobCoordinator leader election has completed. Let's make sure we document somewhere (i.e. depending on where we move this code to).

Reply:
I have added a comment in the code. I looked around to see if there is a way to verify that but couldn't find any.